### ISSUE TITLE
Multi-architecture support

### ISSUE THREAD

**Issue by astefanutti** [apache/camel-k#1238]
At the moment, Camel K binaries and images are available for amd64 architecture only.

Following up on #1215, it would be useful to publish multi-architecture images, so that Camel K can be used on a wider range of architectures. Based on a recent PoC, I've identify the following tasks: 

- Camel K:
    - [ ] Use a multi-architecture default Java base image
    - [ ] Build and publish multi-architecture images: BuildKit could be a good candidate to ease the task of building images for all target architectures, as it automatically picks the right base image and creates the multi-architecture manifest
    - [ ] The https://github.com/scylladb/go-set dependency doesn't compile on 32 bits architectures
- Kaniko:
    - [ ] Multi-architecture based image support: GoogleContainerTools/kaniko#646, GoogleContainerTools/kaniko#980
    - [ ] Multi-architecture _warmer_ and _executor_ images

**Comment by robertonav20** [comment]
Any updates for this task? Can I help with something?

**Comment by squakez** [comment]
@robertonav20 you may have a look at the tasks in the issue description if you're willing to have a look at the implementation. A draft PR with some PoC may be useful to experiment and reason about the pros and cons.

**Comment by robertonav20** [comment]
Hi @squakez,
I build and execute camel k operator and a camel quarkus application inside raspberry.
I used docker buildx and specify the target platform.

I must modify makefile and dockerfile for camel-k-operator and use the correct java base image for camel quarkus application.

I apologize but I haven't strong skill with go language, I try to do my best. Can you give me an advice?

**Comment by squakez** [comment]
I think you can start having a look at the [builder package](https://github.com/apache/camel-k/tree/main/pkg/builder) which is the one in charge to take care for building images and pushing to registry.

**Comment by robertonav20** [comment]
Hi,
i've tried to import BuildKit library like suggest @astefanutti inside the task but there is a problem with dependency

go: finding module for package github.com/go-logr/logr/funcr
github.com/apache/camel-k/pkg/builder imports
        github.com/moby/buildkit/client imports
        go.opentelemetry.io/otel/sdk/trace tested by
        go.opentelemetry.io/otel/sdk/trace.test imports
        github.com/go-logr/logr/funcr: module github.com/go-logr/logr@latest found (v1.2.3, replaced by github.com/go-logr/logr@v0.4.0), but does not contain package github.com/go-logr/logr/funcr

you can find the my code here https://github.com/robertonav20/camel-k/... i will update if i fix the problem

**Comment by robertonav20** [comment]
Hi,
i have an update with this tasks, below the description for each point>
- Use a multi-architecture default Java base image
> I use the graalvm base image in order to use for amd64/arm64 like this
`FROM --platform=$BUILDPLATFORM ghcr.io/graalvm/graalvm-ce:ol8-java11-22.1.0 `

I thinked to use quarkus-mandrel-base-image for arm64 but isn't available for now (https://github.com/quarkusio/quarkus-images/commit/21e25a3c3ae3e3f4a4bd02b1e76f6051c31729c0)

- Build and publish multi-architecture images: BuildKit could be a good candidate to ease the task of building images for all target architectures, as it automatically picks the right base image and creates the multi-architecture manifest
> I modified the Makefile to use BuildKit with correct platform parameter in base of host's architecture
`	
docker buildx rm --all-inactive --force
docker buildx create --append --name builder
docker buildx build --platform=$(IMAGE_TARGET_PLATFORM) -t $(CUSTOM_IMAGE):$(CUSTOM_VERSION) -f build/Dockerfile .
`

For the integration-kit image, I cannot find a solution to build and push a working image because there are a problem dependency (read the previous comment).


So, for the first part can i submit pr?

**Comment by robertonav20** [comment]
It's a pleasure help the community.... For now I try to build an integration kit multi arch with buildah strategy.

It's possible just update buildah version to 1.23.3 because the latest version support arm64 arch... Stay tuned I will give some feedback soon

**Comment by robertonav20** [comment]
As i say previously i build and execute an integration kit in arm64 with buildah, below the modifies:
- Update buildah at 1.23.3 version
- Add the parameter platform
`
bud := []string{
        "buildah",
        "bud",
        "--storage-driver=vfs",
        "--platform",
        platforms.DefaultSpec().OS + "/" + platforms.DefaultSpec().Architecture + "/" + platforms.DefaultSpec().Variant,
        "--pull-always",
        "-f",
        "Dockerfile",
        "-t",
        task.Image,
        ".",
}
`

![image](https://user-images.githubusercontent.com/6991920/169577971-a92cb4cb-d4de-4435-adca-5260e4b8b387.png)


![image](https://user-images.githubusercontent.com/6991920/169577596-20e60a2b-4ed8-4e65-a231-a54c152e8ef3.png)

**Comment by robertonav20** [comment]
@squakez sorry, I don't understand why should be parameterized? 
The integration kit are builded inside kubernetes so with architecture of the node amd64, arm64 etc....  this thing can't be modified by the user.

Can you describe me?

**Comment by squakez** [comment]
Yeah, I am thinking on a generic situation where you have multiple nodes in your cluster, and the operator is running on any non-ARM architecture. Then you have a special node in ARM architecture where you are willing to run the Integrations. In such case, having that value as a parameter would help.

**Comment by robertonav20** [comment]
Ok Great! Thanks!
Anyway you are limited by buildah architecture support, for others solutions it's necessary fix the dependency problem which i describe some comment ago


I will try to add the platform field as attribute of IntegrationPlatform struct, but i don't know these classes and model, can you suggest how to proceed?.

**Comment by squakez** [comment]
You can have a look at some refactoring we've done recently here https://github.com/apache/camel-k/pull/3032 - We've introduced a generic `PublishStrategyOptions` in the IntegrationPlatform that can be used for any strategy extra option (we're using for Kaniko there). I think we could leverage that and from that onward cascade the parameter up to the Buildah Task. Ideally we don't want a specific parameter to slip into an abstract concept (such as the Builder). Also, feel free to create any draft PR and we can discuss over there any idea you have around this or other developments.

**Comment by robertonav20** [comment]
Ok @squakez, i understand your advice but i dont know how to use PublishStrategyOptions.

Can you give me and example to fill that structure?

**Comment by squakez** [comment]
Sure, you can have a look at #3032 where we used that for the "Kaniko" options. We may mimick that and include any "Buildah" option as well.

**Comment by robertonav20** [comment]
Yes i have a look yesterday, below i describe the code:

Inside the package builder i add buildah.go file to manage buildah constants properties
```
package builder

const BuildahArchitecturePlatform = "BuildahArchitecturePlatform"
const BuildahPlatform = "BuildahPlatform"
```

Here you can find the mapping to buildah builder
```
	case v1.IntegrationPlatformBuildPublishStrategyBuildah:
		var architecture string
		var found bool
		if architecture, found = e.Platform.Status.Build.PublishStrategyOptions[builder.BuildahPlatform]; !found {
			architecture = platforms.DefaultSpec().OS + "/" + platforms.DefaultSpec().Architecture + "/" + platforms.DefaultSpec().Variant
		}

		e.BuildTasks = append(e.BuildTasks, v1.Task{Buildah: &v1.BuildahTask{
			BaseTask: v1.BaseTask{
				Name: "buildah",
			},
			PublishTask: v1.PublishTask{
				Architecture: architecture,
				Image:        getImageName(e),
				Registry:     e.Platform.Status.Build.Registry,
			},
			Verbose: t.Verbose,
		}})
```
Then map architecture attribute to buildah command.

I see a difference between https://github.com/apache/camel-k/pull/3032 and this task, that task introduce a command parameter for **install**, but in our case we must manage the architecture of build for integration kit, so it's necessary introduce this parameter inside **run** command

My questions are:
-  Where is the point of the code ? I suppose **run.go** inside **pkg/cmd** and here i can add the new buildah parameter (exmaple: **buildah-platform**)
- The variable `platform` inside `run.go` method `createOrUpdateIntegration` is empty, How to fill it ?
- How to check if that map `PublishStrategyOptions` contains that parameter?

I'm hope you understand my questions.

Thanks in advance

**Comment by squakez** [comment]
Any build configuration should be included into the IntegrationPlatform, not exposed in the `kamel run` command. I think you can extend the BuildahTask to include the new needed parameters
https://github.com/apache/camel-k/blob/fe997b9ac1297884840ffb42bb41302bb5040ee5/pkg/apis/camel/v1/build_types.go#L107
and passing them from IntegrationPlatform.PublishStrategyOptions as we're doing for Kaniko settings. Once they are in the BuildahTask, you should be able to use them.

**Comment by robertonav20** [comment]
Ok thanks, i understand your advice and i fix the code just like this.

```
type BuildahTask struct {
	BaseTask    `json:",inline"`
	PublishTask `json:",inline"`
	// The platform of build image
	Platform string `json:"baseImage,omitempty"`
	// log more information
	Verbose *bool `json:"verbose,omitempty"`
}

case v1.IntegrationPlatformBuildPublishStrategyBuildah:
	var platform string
	var found bool
	if platform, found = e.Platform.Status.Build.PublishStrategyOptions[builder.BuildahPlatform]; !found {
		platform = platforms.DefaultSpec().OS + "/" + platforms.DefaultSpec().Architecture + "/" + platforms.DefaultSpec().Variant
	}

	e.BuildTasks = append(e.BuildTasks, v1.Task{Buildah: &v1.BuildahTask{
		Platform: platform,
		BaseTask: v1.BaseTask{
			Name: "buildah",
		},
		PublishTask: v1.PublishTask{
			Image:    getImageName(e),
			Registry: e.Platform.Status.Build.Registry,
		},
		Verbose: t.Verbose,
	}})
```
In this way you can manage the platform of image if you want, just fill `PublishStrategyOptions` with the correct paramater from outside.

But, i don't understand how to works `PublishStrategyOptions`, for kaniko i found this code inside `install.go`

```
kanikoBuildCacheFlag := cobraCmd.Flags().Lookup("kaniko-build-cache")
if kanikoBuildCacheFlag.Changed {
	platform.Spec.Build.PublishStrategyOptions[builder.KanikoBuildCacheEnabled] = strconv.FormatBool(o.KanikoBuildCache)
}
```

If i undertstand well, with install command you can pass `kaniko-build-cache` parameter and to set the PublishStrategyOptions, right?
There is another way to fill `PublishStrategyOptions`?
How do you thing to manage the `platform` parameter for buildah?

Thanks in advance! 😄

**Comment by squakez** [comment]
Nice, that fits perfectly into the design. I think those parameters must be provided by the user editing the IntegrationPlatform (manually) ie `kubectl edit ip`. A nice addition would be to transform that `kaniko-build-cache` into something generic, so we can manage any PublishStrategyOptions option. I'm creating a follow up issue for that. For now, I guess that the manual editing of the IntegrationPlatform would be enough for the scope.

**Comment by robertonav20** [comment]
Ok great! So you leave to the user edit the manifest of integration platform under Camel K CRD.

Can i submit a PR?

**Comment by squakez** [comment]
Sure, it's more than welcome.

**Comment by robertonav20** [comment]
You can find here https://github.com/apache/camel-k/pull/3309 the pr

**Comment by barius** [comment]
I'm able to build a arm64 camel-k-operator image by first creating a arm64 quarkus-mandel image, then base camel-k on this.
However, integration kits built in this arm64 operator is still amd64/x64.
I'm running Docker Desktop on an M1 Macbook, camel-k 1.9.1.

Looked into camel-k builds and saw the Spectrum builder. Don't know if it supports arm64.
Is there any way I can get a arm64 image? Maybe after the above PR is merged? Or is there anyway I can help with that?

**Comment by robertonav20** [comment]
Hi @barius to build integration kits arm64 it's necessary use buildah as build-strategy you can follow the pr open from this ticket

**Comment by barius** [comment]
> 

Okay I'll try it! Good work!

**Comment by barius** [comment]
Confirm that the PR works!
Now my integrations are arm64, yeah! Saw ~4x performance on my mac :)
I'll still using my local arm64 quarkus-mandrel image though, didn't find the manual command to build kamel operator to support arm architecture :(

**Comment by robertonav20** [comment]
Ok great you can find the full guide here https://github.com/robertonav20/camel-k/blob/main/docs/modules/ROOT/pages/configuration/multi-architecture.adoc

**Comment by barius** [comment]
Ah I see, great! However it uses Graal. Don't know what the difference between mandrel and graal so I'm waiting for the official release :) For now everything works for me and I don't have much time testing it out :P

**Comment by squakez** [comment]
Partially available in #3309

**Comment by robertonav20** [comment]
Can we close this issue?

**Comment by squakez** [comment]
No, I think that although partially available with your development, this issue need some more work to be fully completed.

**Comment by barius** [comment]
Although Buildah is perfectly fine with regards to building kit images, when Istio auto-injection is enabled, the camel-k builder always wait for the Istio container to complete (which will never happen), and triggers the default 5 min timeout (though Buildah finished in around 1~2 min and after the timeout the build still succeeds).

Don't know if this is the expected behavior, or if I should open another issue with it.

**Comment by squakez** [comment]
@barius please better open a new issue and which version is affected (I suppose the nightly build).

**Comment by barius** [comment]
@squakez yes pls take a look #3408

**Comment by robertonav20** [comment]
@squakez now the is available the multi image of `ubi-quarkus-mandrel` and `ubi-quarkus-micro-image`, so you can update the dockerfile to support arm64 natively then delete the new file `Dockerfile.arch`.

Here the quay.io repository
`podman pull quay.io/quarkus/quarkus-micro-image:2.0`
`podman pull quay.io/quarkus/ubi-quarkus-mandrel:22.0.0.2-Final-java11-arm64`

Ref. to the closed issue https://github.com/quarkusio/quarkus-images/issues/83

**Comment by squakez** [comment]
Cool! Thanks for the heads up. Feel free to have a look yourself while this is not picked up for development.

**Comment by tobiasoort** [comment]
> Can we close this issue?

Not really. The current state on Apple M1/M2 is that you can follow the how-to, but 'kamel install' will install an amd64 image based pod (image apache/camel-k from dockerhub is only linux/amd64) into your m1-powered k8s env, which obviously doesn't work. 

Having true multi-arch images would be great for everyone! This means changing the build pipeline that builds and pushes images over to dockerhub.

I'd love to help where possible. I'm quite alright with docker/kubernetes but my Java is really rusty - especially modern buildchains. I'll see where I can help and go through this thread.

For future googlers, the following error in your pod's logs:
`standard_init_linux.go:228: exec user process caused: exec format error`
means you're running the wrong arch (in this specific case, running an amd64 image on arm64).

**Comment by squakez** [comment]
> Edit: It seems I was being captain obvious. Is there anything I can help with?
> 
> For future googlers, the following error in your pod's logs: `standard_init_linux.go:228: exec user process caused: exec format error` means you're running the wrong arch (in this specific case, running an amd64 image on arm64).

I guess the last message from @robertonav20 could be a good point from where to continue with the development of this issue.

**Comment by tobiasoort** [comment]
So, based on the assumptions:
- ubi-quarkus-mandrel v22.0.0.2-Final-java11 is the version to target (v21 doesnt have arm builds)
- hosting the application in the new quarkus container is enough - the Maven build doesnt need to change (yay jvm?)
- I know how buildx works

I have submitted https://github.com/apache/camel-k/pull/3820

**Comment by tobiasoort** [comment]
> * I know how buildx works

Foreshadowing quote was exactly that. I'm afraid i'll break the build because with multi-arch images we can't rely on the local image store (see https://github.com/docker/buildx/issues/59). I've worked around it and hope that none of the processes rely on 'the image just being there'.

Also, I see a 'docker build' step for a 'bundle'. What is that 'bundle', and do I need to fix that too?

**Comment by tobiasoort** [comment]
Question for whoever is reading this: when pushing the image to the kind-registry, the following error occurs:

```
#22 ERROR: failed to do request: Head "https://kind-registry:5000/v2/apache/camel-k/blobs/sha256:f7115ccb0c19b17f613dcdab9c9ff5b95451b8af76ff62c4b482edb1543b1436": dial tcp: lookup kind-registry on 168.63.129.16:53: no such host
```
Is kind-registry on and available by default? Or is this a proxy service that only exists/works in the original repo?

**Comment by tobiasoort** [comment]
So, the issue is:

- The `IMAGE_NAME` is set in the `secrets.E2E_CLUSTER_CONFIG` secret to `kind-registry:5000/apache/camel-k` (I think, its a secret!)
- The `IMAGE_NAME` contains the hostname `kind-registry`
- This hostname is not resolving to 127.0.0.1 and thus it errors out.

How can we validate this? Can someone share the outline of the `secrets.E2E_CLUSTER_CONFIG` so I can override it in my own fork?

**Comment by tobiasoort** [comment]
Integration tests are failing due to:

```
 Error: p.go:274:     > {"level":"error","ts":1668698213.7254713,"logger":"camel-k.maven.build","msg":"\t[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.IllegalStateException: Out of date version of GraalVM detected: native-image 22.0.0.2-Final Mandrel Distribution (Java Version 11.0.14+9). Quarkus currently supports 22.3.0. Please upgrade GraalVM to this version."
```

I'll see what else collapses when we push GraalVM to that version :)

**Comment by squakez** [comment]
> g":"\t[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.IllegalStateException: Out of date version of GraalVM detected: native-image 22.0.0.2-Final Mandrel Distribution (Java Version 11.0.14+9). Quarkus currently supports 22.3.0. Please upgrade GraalVM to this version."

Known issue #3817

**Comment by tobiasoort** [comment]
Current status: ready for polishing. All pipeline errors are due to #3817 or timeouts that are also already known. Next up: parameterize the build archs.

**Comment by tobiasoort** [comment]
Current status: retracted the PR because the build is failing at a weird point (unittests of the client, even before the Operator is being built...). Will fiddle in my own fork and create a new and clean one once it all works.

**Comment by squakez** [comment]
As we're planning to rework the way we build the images, I think we should take the opportunity to include the multi-architecture support as well.

**Comment by squakez** [comment]
I am closing this issue in favor of #4148 and #4149 where we can track separately the evolution of Operator images and Integration images.
